java-checks/src/main/java/org/sonar/java/checks/NullPointerCheck.java

```diff
@@ -0,0 +1,147 @@
+    this.context = context;
+    if (context.getSemanticModel() != null) {
+      expressionVisitor.scan(context.getTree());
+    }
+  }
+
+  /** this visitor invokes issueVisitor for the base expressions of array access, member selectm and method invocation expressions. */
+  private class ExpressionVisitor extends BaseTreeVisitor {
+    @Override
+    public void scan(@Nullable Tree tree) {
+      super.scan(tree);
+    }
+
+    @Override
+    public void visitArrayAccessExpression(ArrayAccessExpressionTree tree) {
+      issueVisitor.scan(tree.expression());
+      super.visitArrayAccessExpression(tree);
+    }
+
+    @Override
+    public void visitMemberSelectExpression(MemberSelectExpressionTree tree) {
+      issueVisitor.scan(tree.expression());
+      super.visitMemberSelectExpression(tree);
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
+      scan(tree.typeArguments());
+      for (ExpressionTree argument : tree.arguments()) {
+        issueVisitor.scan(argument);
+      }
+      scan(tree.methodSelect());
+    }
+  }
+
+  /** this visitor raises an issue if it visits an identifier, method invocation or member select whose symbol can be null. */
+  private class IssueVisitor extends BaseTreeVisitor {
+    @Override
+    public void scan(@Nullable Tree tree) {
+      super.scan(tree);
+    }
+
+    @Override
+    public void visitIdentifier(IdentifierTree tree) {
+      Symbol symbol = ((SemanticModel) context.getSemanticModel()).getReference((IdentifierTree) tree);
+      if (isAnnotationNull(symbol)) {
+        context.addIssue(tree, RULE_KEY, String.format("Field '%s' can be null.", symbol.getName()));
+      }
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
+      Symbol symbol = ((MethodInvocationTreeImpl) tree).getSymbol();
+      if (isAnnotationNull(symbol)) {
+        context.addIssue(tree, RULE_KEY, String.format("Value returned by method '%s' can be null.", symbol.getName()));
+      }
+      /* continues processing using the expression visitor (base expressions are not necessarily nullable). */
+      expressionVisitor.visitMethodInvocation(tree);
+    }
+
+    @Override
+    public void visitMemberSelectExpression(MemberSelectExpressionTree tree) {
+      IdentifierTree identifierTree = ((MemberSelectExpressionTree) tree).identifier();
+      Symbol symbol = ((SemanticModel) context.getSemanticModel()).getReference(identifierTree);
+      if (isAnnotationNull(symbol)) {
+        context.addIssue(identifierTree, RULE_KEY, String.format("Field '%s' can be null.", symbol.getName()));
+      }
+      /* continues processing using the expression visitor (base expressions are not necessarily nullable). */
+      expressionVisitor.visitMemberSelectExpression(tree);
+    }
+
+  }
+
+  /** returns true if the symbol is annotated with CheckForNull or Nullable. */
+  private boolean isAnnotationNull(Symbol symbol) {
```

Comment:
Ok. name will be changed anyway because we don't want to rely only on annotations.